### PR TITLE
some final inference fixes for custom number of codebooks

### PR DESCRIPTION
**Pull request by paarthneekhara**

### CODE REVIEW

File: examples/nlp/language_modeling/speechllm_multitask_dataprep.py [hunk 1]
```diff
@@ -88,7 +88,7 @@ def __init__(
             self.add_context_records_to_manifest()
         
         self.base_data_dir = get_base_dir([item["audio_filepath"] for item in self.data])
-        self.filter_invalid_records()
+        # self.filter_invalid_records()
```
Comment (blisc) [inline examples/nlp/language_modeling/speechllm_multitask_dataprep.py:91]: Why is this removed?

File: nemo/collections/nlp/models/language_modeling/megatron_t5_speechlm_model.py [hunk 2]
```diff
@@ -1108,12 +1116,15 @@ def build_virtual_prompt_dataset(
...
+            add_special_tokens_to_only_first_codebook=self.cfg.data.get('add_special_tokens_to_only_first_codebook', False),
         )
 
         rank = parallel_state.get_data_parallel_rank()
         world_size = parallel_state.get_data_parallel_world_size()
+        # generate a random seed to shuffle data differently in each run
+        seed = int(time.time())
         sampler = torch.utils.data.distributed.DistributedSampler(
-            dataset, num_replicas=world_size, rank=rank, shuffle=shuffle, seed=self.cfg.seed
+            dataset, num_replicas=world_size, rank=rank, shuffle=shuffle, seed=seed
```
Comment (blisc) [inline nemo/collections/nlp/models/language_modeling/megatron_t5_speechlm_model.py]: This is not a good idea. It should be fixed across different ranks. See https://pytorch.org/docs/stable/data.html#torch.utils.data.distributed.DistributedSampler

File: nemo/collections/nlp/models/language_modeling/megatron_t5_speechlm_model.py [hunk 3]
```diff
@@ -323,7 +324,13 @@ def load_frozen_model(self, cfg, trainer):
             t5_cfg.micro_batch_size = cfg.get('micro_batch_size', 4)
             t5_cfg.global_batch_size = cfg.get('global_batch_size', 4)
             t5_cfg.precision = trainer.precision
-            t5_cfg.tokenizer.num_sentinel_tokens = 39184 - 29056  # cfg.num_speech_tokens 39168
+            # TODO: Remove hardcoding
+            if cfg.data.get('speech_offset', 30000) > 250000:
```
Comment (blisc) [inline nemo/collections/nlp/models/language_modeling/megatron_t5_speechlm_model.py:329]: Do you know what num_sentinel_tokens is used in?

**Review by blisc — CHANGES_REQUESTED** [review]

File: examples/nlp/language_modeling/speechllm_multitask_dataprep.py (hunk 1 above)
Comment (paarthneekhara) [reply examples/nlp/language_modeling/speechllm_multitask_dataprep.py:91]: The filter_invalid_records function was added because some LibriTTS audio files (5-10) on my machine were corrupt. But it takes a while to filter invalid records for other datasets (HiFiTTS, Riva etc) which don't have this issue. So I have commented it for now.

File: nemo/collections/nlp/models/language_modeling/megatron_t5_speechlm_model.py (hunk 2 above)
Comment (paarthneekhara) [reply nemo/collections/nlp/models/language_modeling/megatron_t5_speechlm_model.py]: I see. Sure, let me remove this.

File: nemo/collections/nlp/models/language_modeling/megatron_t5_speechlm_model.py (hunk 3 above)
Comment (paarthneekhara) [reply nemo/collections/nlp/models/language_modeling/megatron_t5_speechlm_model.py:329]: Yes, it affects the size of the encoder and decoder embedding matrices, so it raises an error that the size of embedding matrix in the model and checkpoint are not the same.

**Review by blisc — APPROVED** [review]
